_posts/2018-05-23-gracefully_clean_up_in_grpc_junit_tests.md

```diff
@@ -0,0 +1,115 @@
+
+<!--more-->
+
+So cleanup, cleanup, cleanup...\
+and fail the test if any cleanup is not successful.
+
+A typical example is
+
+```java
+public class MyTest {
+  private Server server;
+  private ManagedChannel channel;
+  ...
+  @After
+  public void tearDown() InterruptedException {
+    // assume channel and server are not null
+    channel.shutdownNow();
+    server.shutdownNow();
+    // fail the test if cleanup is not successful
+    assert channel.awaitTermination(5, TimeUnit.SECONDS) : "channel failed to shutdown";
+    assert server.awaitTermination(5, TimeUnit.SECONDS) : "server failed to shutdown";
+  }
+  ...
+}
+```
+
+or to be more graceful
+
+```java
+public class MyTest {
+  private Server server;
+  private ManagedChannel channel;
+  ...
+  @After
+  public void tearDown() throws InterruptedException {
+    // assume channel and server are not null
+    channel.shutdown();
+    server.shutdown();
+    // fail the test if cannot gracefully shutdown
+    try {
+      assert channel.awaitTermination(5, TimeUnit.SECONDS) : "channel cannot be gracefully shutdown";
+      assert server.awaitTermination(5, TimeUnit.SECONDS) : "server cannot be gracefully shutdown";
+    } finally {
+      channel.shutdownNow();
+      server.shutdownNow();
+    }
+  }
+  ...
+}
+```
+
+But that's tedious. You need to write the shutdown boilerplate by yourself. So gRPC testing library introduced something to make it less tedious.
+
+A JUnit rule [`GrpcServerRule`][GrpcServerRule] was first introduced to eliminate the shutdown boilerplate. With this rule, it creates an In-Process server and channel at the beginning of the test and shuts them down at the end of test automatically. But it is inflexible and too restrictive in the sense that it does not support transports other than In-Process transports, multiple channels to the server, custom channel or server builder options, and configuration inside individual test methods.
+
+A more preferable JUnit rule [`GrpcCleanupRule`][GrpcCleanupRule] was introduced (and will be available from gRPC release v1.13.0), which also eliminates the shutdown boilerplate. Unlike `GrpcServerRule`, `GrpcCleanupRule` does not create any server or channel automatically at all. Users should create and start the server by themselves, and create channels by themselves, just as in plain tests. With this rule, users just need to register every resource (channel or server) that needs to be shut down at the end of test, and the rule will then shut them down gracefully automatically.
+
```

Comment:
A more flexible JUnit rule [`GrpcCleanupRule`][GrpcCleanupRule] was introduced from gRPC release v1.13.0, which also eliminates the shutdown boilerplate. However unlike `GrpcServerRule`, `GrpcCleanupRule` does not create any server or channel automatically at all. Users create and start the server by themselves, and create channels by themselves, just as in plain tests. With this rule, users just need to register every resource (channel or server) that needs to be shut down at the end of test, and the rule will then shut them down gracefully automatically.

Reply:
Link doesn't work again.

Reply:
Done. I checked the link in preview, seems working.